ADALiOS/ADALiOS/ADAuthenticationResult+Internal.m

```diff
@@ -109,32 +109,54 @@ +(ADAuthenticationResult*) resultFromCancellation
 
 +(ADAuthenticationResult*) resultFromBrokerResponse: (NSDictionary*) response
 {
-    ADAuthenticationError* error;
-    ADAuthenticationResult* result;
+    ADAuthenticationResult* result = nil;
     ADTokenCacheStoreItem* item = nil;
-    if([response valueForKey:OAUTH2_ERROR_DESCRIPTION])
-    {
-        error = [ADAuthenticationError errorFromNSError:[NSError errorWithDomain:ADBrokerResponseErrorDomain code:0 userInfo:nil] errorDetails:[response valueForKey:OAUTH2_ERROR_DESCRIPTION]];
-    }
-    else
+    
+    if(!response || [response valueForKey:OAUTH2_ERROR_DESCRIPTION])
     {
-        item = [ADTokenCacheStoreItem new];
-        item.authority =  [response valueForKey:OAUTH2_AUTHORITY];
-        item.resource = [response valueForKey:OAUTH2_RESOURCE];
-        item.clientId = [response valueForKey:OAUTH2_CLIENT_ID];
-        item.accessToken = [response valueForKey:OAUTH2_ACCESS_TOKEN];
-        item.refreshToken = [response valueForKey:OAUTH2_REFRESH_TOKEN];
-        if([response valueForKey:OAUTH2_ID_TOKEN])
+        ADAuthenticationError* error = nil;
+        NSString* errorDetails = nil;
+        NSInteger errorCode = 0;
+        if (response)
         {
-            ADUserInformation* info = [ADUserInformation userInformationWithIdToken:[response valueForKey:OAUTH2_ID_TOKEN] error:&error];
-            if(!error)
+            errorDetails = [response valueForKey:OAUTH2_ERROR_DESCRIPTION];
+            errorCode = [[response valueForKey:@"error_code"] integerValue];
+            
+            if (!errorDetails)
             {
-                item.userInformation = info;
+                errorDetails = @"Broker did not provide any details";
             }
         }
+        else
+        {
+            errorDetails = @"No broker response received.";
+        }
+        
+        error = [ADAuthenticationError errorFromNSError:[NSError errorWithDomain:ADBrokerResponseErrorDomain code:errorCode userInfo:nil] errorDetails:errorDetails];
+        
+        return [ADAuthenticationResult resultFromError:error];
     }
     
-    result.tokenCacheStoreItem.accessTokenType = @"Bearer";
+    ADUserInformation* info = nil;
+    if([response valueForKey:OAUTH2_ID_TOKEN])
+    {
+        ADAuthenticationError* error = nil;
+        info = [ADUserInformation userInformationWithIdToken:[response valueForKey:OAUTH2_ID_TOKEN] error:&error];
+        if (error)
```

Comment:
I ended up copying over the existing behavior on the non-broker code path where if idtoken fails to decode it'll just continue onwards. Whether or not that's the right behavior for either case is up for discussion. :+1: